Rocksniffer: Playthrough Logging and Game State Detection

PoizenJam opened the pull request:
# Improved Game State Detection and Playthrough Logging

This PR collects the RockSniffer_PJ fork work since it diverged from upstream, bringing it in like with the main fork's structure and adds new capability such as: game-state detection, memory reverse-engineering, playthrough logging, and the addon overlays. This PR depends on RockSnifferLib PR changes in the companion PR #16

The following document is organized by module change rather than commits, since the 39 commits are mostly incremental hotfixes within each area. Where review feedback in this PR later revised a feature, the text below describes the current state, not the original version.

## Game state and game stage detection

Two output files expose game state for OBS scene switching and other tooling, both set via format tokens in `config/output.json`. `game_state.txt` (`%GAME_STATE%`) carries the high-level SnifferState (IN_MENUS, SONG_STARTING, SONG_PLAYING, SONG_PAUSED, SONG_ENDING). `game_stage.txt` (`%GAME_STAGE%`) carries Rocksmith's own internal gameStage string read from memory.

The gameStage reader was moved off a fragile four-hop pointer chain into a transient UI struct, which returned garbage in Score Attack menus and never saw pause stages, and onto a direct read of the `.data`-section buffer at `Rocksmith2014.exe+0xF5F7C9` (Remastered). That fixed the junk values in SA pre-song menus and surfaced three stages the old reader dropped (las_pause, nsp_pause, nsp_tuner). One piece of engine behavior is documented in the code so nobody "fixes" it later: Rocksmith does not update this cell on pause-to-resume or pause-to-restart, so play/pause state **has to come from SnifferState**, not from parsing the gameStage string.

Several cases that used to leave the state machine stuck now recover: attaching to an already-running Rocksmith, Rocksmith restarting mid-session, and backing out of a loading screen before gameplay starts. The IN_MENUS to SONG_SELECTED transition is gated on gameStage rather than on `songTimer != 0` alone, which stops transient memory garbage at startup from marching the state forward, and SONG_STARTING has an escape back to IN_MENUS. Both guards exclude the sticky `*_pause` stages.

## Pause and resume detection

Pause detection started as a timer-stall heuristic with multi-poll latency, false positives near the end of a song, and no coverage of Guitarcade. It now reads Rocksmith's pause-menu mode byte directly from a `.data` static at `Rocksmith2014.exe+0xF5F5FC` (Remastered), which gives first-poll pause and resume across all song modes and the Guitarcade minigames. The tuner-from-pause case is handled without special casing: the engine moves 2, 1, 2 between the pause menu and the tuner without ever dropping to 0, so the paused state holds. Pause-exit compares against a timer snapshot taken at pause entry, so quitting from the pause menu doesn't flicker through SONG_PLAYING on the way to IN_MENUS. Cross-mode validated, and confirmed to survive relaunch as a true static.

The old stall-counter machinery (STALL_THRESHOLD, stallCount, the end-of-song guard) is gone.

## Memory reverse-engineering

Three memory reads were added, each documented in the code with its pointer chain, value mapping, edition-shift caveats, and validation status. currentPath is the selected instrument path (Lead, Rhythm, Bass) from a stable byte pointer that populates from launch and holds in every game stage including Nonstop Play; it's the primary arrangement-resolution fallback when the arrangement-hash pointer isn't available. pauseMenuMode is the blocking-overlay byte described above. PLAY_arrID is the loaded arrangement GUID, read through a five-deep chain that populates in Nonstop Play where the legacy arrangement-hash chain does not, which closed the long-standing Nonstop arrangement-ID gap. The arrangement-ID read dispatches on gameStage: PLAY_arrID for LaS and Nonstop, legacy arrangement-hash for Score Attack and menus.

As I was working with the latest version of Rocksmith 2014, Beta and Learn-And-Play offsets are back-derived from the `+0x3080` / `+0x4080` shift convention used across the memory map and are marked unverified.

## Mode classification

The `mode` field used to be set from note-data pointer resolution, which classified Nonstop as LEARNASONG (it reuses the LaS note-data subsystem) and left every menu state UNKNOWN. It's now derived from gameStage through a single `DeriveModeFromGameStage` classifier. The RSMode enum grew from 4 values to 9 (adding NONSTOPPLAY, GUITARCADE, SESSION, LESSONS, MENU), with the original integer values 0 through 3 unchanged. The bare `tuner` stage has no mode of its own, so it keeps the previous poll's mode. There's now exactly one place that writes `mode`.

## Playthrough history (SQLite and CSV)

Each playthrough logs to a SQLite database and/or CSV with full metadata, configured in `config/history.json`. A record holds three timestamps (metadata load, gameplay start, gameplay end, which lets you line records up against a stream recording), song metadata, arrangement path and tuning, performance stats (notes, hits, misses, streak, accuracy), completion and pause flags, and Score Attack stats where they apply.

The completed flag on force-end paths is decided by deterministic signals in `DetermineCompletedForForceEnd` rather than the old `maxTime >= songLength - 0.5s` guess. It catches pause-driven exits, no-input boots, and Score Attack three-strike fails, and it correctly marks natural completions that missed the SONG_ENDING transition because of the poll cadence.

Nonstop Play is tracked in full, unblocked once PLAY_arrID made the arrangement data trustworthy. Multiplayer is held out of history until there's real multi-user support, since MP shares gameStages with the single-player modes and can't be disambiguated from gameStage alone.

## Output format tokens

New tokens for `output.json`: `%CURRENT_PATH%` (Lead/Rhythm/Bass), `%TUNING%` (resolved tuning name), and `%CURRENT_SECTION%` / `%CURRENT_PHRASE%` (chart-authored names at the current song timer). Each writes an empty string when it can't resolve, so the OBS text source hides on its own. The new default output entries (path.txt and the section/phrase files) merge into an existing config automatically.

## Addon overlays

Nonstop Play now resolves correctly in the overlays via currentPath, so sections, phrases, and tuning render instead of falling through to the old broken default-path heuristic.

The vocal overlay was rewritten for karaoke-style display: DOM-tracked line elements that slide through slot positions, matching the in-game behavior for line breaks, highlighting, and marking sung lyrics, with rocksmith-style, twitch, and kick themes.

Two mode-1 results-view fixes: a minimum hold so the post-song comparison isn't wiped the instant the Nonstop songID flips to the next queued song, and an idempotent finalize so the "% vs previous best" line reads the real final-section accuracy instead of a placeholder.

The arrangement-resolution and results-hold logic that was copy-pasted across the addons now lives once in `sniffer-poller.js` (`resolveArrangement` and `enableResultsHold`), and the addons call into it.

Fonts are bundled locally instead of pulled from Google Fonts, which avoids sending user IPs to Google (GDPR compliance issue). They were moved to `addons/_deps/fonts/` and are served with correct content types.

## Tuning detection

The tuning dictionary was expanded for rare and exotic tunings and to paper over common CDLC metadata mistakes, like bass charts that don't set the B and high-e offsets. If nothing in the dictionary matches, it falls back to the raw per-string tuning instead of 'Custom Tuning'. Full support for display of non-standard Pitch and/or Capo information in add ons and history.

## Configuration changes

Playthrough history settings moved to their own `config/history.json` (enableSqliteHistory, sqliteHistoryPath, enableCsvHistory, csvHistoryPath). This is a clean break, so existing users re-set those flags in the new file. Output files are also cached and only rewritten when their content changes.

## Breaking changes (readout JSON)

Third-party addons that read the readout JSON need one-line updates. Every bundled addon is already updated.

- `mode` serializes as its name ("SCOREATTACK") instead of an integer
- `currentState` serializes as its name ("SONG_PAUSED") instead of an integer
- `pauseMenuMode` serializes as "None" / "SubOverlay" / "TopOverlay" instead of 0/1/2
- `currentPath` serializes as "Lead" / "Rhythm" / "Bass" / "Unknown" (the RSPath enum values equal the raw bytes, so the byte is recoverable by a cast)

## Compatibility and validation

Verified on Rocksmith 2014 Remastered and tested across Learn-A-Song, Score Attack, Nonstop Play, and Guitarcade. Beta and LaP memory offsets are back-derived and unverified. Existing playthrough_history rows are untouched; only new rows use the corrected classifications.

## Acknowledgments

Thanks to kokolihapihvi for cross-referencing the original reverse-engineering notes, which narrowed the gameStage and pauseMenuMode searches.

PoizenJam committed:
-Add gameStage as game_stage.txt in output directory

PoizenJam committed:
-Added a tracker for "game_state" to the output folder OutputSettings.cs
-Added line to update and write "gameState" variable in Program.cs
-commented out the *sniff sniff* line
-Increased OutputDetails write rate to 100ms

PoizenJam committed:
-Updated visual studio version

PoizenJam committed:
-Change referenced RockSnifferLib.git to PoizenJam Fork.
-Updated Readme

PoizenJam committed:
-Update submodule reference

PoizenJam committed:
Revert "-Change referenced RockSnifferLib.git to PoizenJam Fork."

This reverts commit a70d98e62018f9bd2a47e561f3d5702306262a36.

PoizenJam committed:
Revert "-Added a tracker for "game_state" to the output folder OutputSettings.cs"

This reverts commit c156c3d43d336c2f9aa182f8b6d9a44f89bf9d71.

PoizenJam committed:
-Trying to revert screwups :D

PoizenJam committed:
Merge branch 'master' of https://github.com/PoizenJam/RockSniffer_PJ

PoizenJam committed:
-error fixing

PoizenJam committed:
-attempt 2 at updating submodule

PoizenJam committed:
Update submodule to latest on RockSnifferLib

PoizenJam committed:
-Updated Readme

PoizenJam committed:
-Fixed formatting of readme

PoizenJam committed:
-and again

PoizenJam committed:
-and again

PoizenJam committed:
-Added distinct version number

PoizenJam committed:
Updated version naming convention

PoizenJam committed:
Rocksniffer_PJ 0.4.2 is a small feature update:
-If tuning tuning match in dictionary is not found, it will fall back on raw string tuning; i.e., if 'E Standard' was not found, report 'EADGBE' instead of 'Custom Tuning'
-If using bass, tuning name will be trimmed to four characters

PoizenJam committed:
Rocksniffer_PJ 0.4.2 is a small feature update:
-If tuning tuning match in dictionary is not found, it will fall back on raw string tuning; i.e., if 'E Standard' was not found, report 'EADGBE' instead of 'Custom Tuning'
-If using bass, tuning name will be trimmed to four characters

PoizenJam committed:
Minor bug fixes related to tuning dictionary and detection

PoizenJam committed:
Minor bug fix for A220 tunings

PoizenJam committed:
Increment version number

PoizenJam committed:
Fixing bugs in various overlays that affect the height of the generated phrase/section meters. Specifically, when switching paths, the difficulty bars were previously either too large or too small.

PoizenJam committed:
Deleted redundant 'script - Copy.js' from Arcade_v1_LaS

PoizenJam committed:
Major update:
-Brought RockSniffer_PJ in line with most recent upgrades in main Rocksniffer (as of 2025-11-24)
-Migrated log print functionality from RockSnifferLib to main RockSniffer repo
-Added config enable-able SQL and CSV playthrough history options via output.json

PoizenJam committed:
Update submodules to latest versions

PoizenJam committed:
Re-add deleted addons!!!

PoizenJam committed:
Test

PoizenJam committed:
Update RockSnifferLib submodule to latest

PoizenJam committed:
Added path and tuning information to playthroughHistory tracking.

PoizenJam committed:
Fix ci.yml

PoizenJam committed:
Update RockSnifferLib submodule to latest

PoizenJam committed:
update yml to align with net8.0 update

PoizenJam committed:
Yet Another YML update fix because I'm bad at github.

PoizenJam committed:
Improve Rocksniffer Vocal module to better cohere with Rocksmith's built in style- lyrics now highlight and display exactly as in game.

PoizenJam committed:
Minor update- make sure accuracy is rounded in CSV and DB!!!!

PoizenJam committed:
-Update game_state detection
-Add score attack mode to PlaythroughHistory tracking, complete with relevant statistics tracking.

PoizenJam committed:
Update submodules to latest versions

PoizenJam committed:
Update RockSnifferLib submodule - pause detection fix

PoizenJam committed:
Update RockSnifferLib submodule - pause detection fix

PoizenJam committed:
Fix duplicate history logging, update RockSnifferLib submodule

PoizenJam committed:
Update README for RockSniffer_PJ enhancements

RockSniffer_PJ is an enhanced fork of RockSniffer with features like pause detection, structured logging, and Score Attack support for improved gameplay tracking.

PoizenJam committed:
Update RockSnifferLib submodule - custom tuning display

PoizenJam committed:
-Updated Addons, specifically Vocals

PoizenJam committed:
-Increment version number from original Rocksniffer repo.

PoizenJam committed:
Enhance README with Score Attack and tuning improvements

Added Score Attack support details and improved tuning dictionary and vocal overlay sections.

PoizenJam committed:
Rename project from RockSniffer_PJ to RockSniffer

Updated README to reflect the project name change from RockSniffer_PJ to RockSniffer.

PoizenJam committed:
Update credits section for clarity

PoizenJam committed:
Revise Readme

Friedfisch commented on addons/vocals/kick.css:
not save regarding european laws, GDPR

PoizenJam replied:
Could you elaborate/clarify?

Friedfisch replied:
due to european laws / GDPR linking google fonts is not allowed. you can download those for free and embed them, so far that is allowed.

PoizenJam replied:
This should be resolved in the latest commit. Fonts now distributed and loaded locally.

PoizenJam committed:
-Update CSS files and font assignments in vocal addon overlay to be Euro / GDPR compliant (i.e., font loaded locally rather than imported from URL)

PoizenJam committed:
Merge branch 'master' of https://github.com/PoizenJam/RockSniffer

Friedfisch commented on addons/current_song_v4/current_song.html:
One more font thing :)

PoizenJam replied:
Fixed in recent commit, including in other files not mentioned ;)

Friedfisch commented on addons/vocals/vocals.html:
```diff
@@ -5,14 +5,12 @@
+		<!-- you can choose style.css (legacy), rocksmith-style.css (Default), twitch.css or kick.css-->
 		<link rel="preconnect" href="https://fonts.googleapis.com">
 		<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
 		<link href="https://fonts.googleapis.com/css2?family=Open+Sans:ital,wght@0,300..800;1,300..800&display=swap" rel="stylesheet">
```
Font again

PoizenJam replied:
Fixed in recent commit, including in other files not mentioned ;)

Friedfisch commented on RockSniffer/Logging/EventLogger.cs:
```diff
@@ -0,0 +1,134 @@
+    /// </summary>
+    public static class EventLogger
+    {
+        public static string eventLogMode = "disabled"; // "disabled", "enabled", "legacy"
```
Maybe use an enum for that? Currently the comment options do not match the comment in OutputSettings.cs L38++

PoizenJam replied:
Good point! Implemented in latest commit.

Friedfisch commented on RockSniffer/RPC/DiscordRPCHandler.cs:
This looks like a bugfix from an earlier push? Not sure what is going on here but "-readout.songTimer" in both arguments seems to be replacable by not even touching songTimer? Or is this to cover "ingame pause"?

PoizenJam replied:
I am not responsible for this code, but I _believe_ the purpose here is for calculating multiple different kinds of useful songTimer values? i.e., start time, current time vs. remaining time, end time? Not confident enough to touch this or mark as resolved.

kokolihapihvi replied:
I believe it is because Discord wants the timestamps "when the song started" - "when the song will end", so it is calculating those from the current UTC time and the song timer.

Friedfisch commented on RockSniffer/History/PlaythroughHistory.cs:
```diff
@@ -0,0 +1,337 @@
+                    artist_name TEXT,
+                    album_name TEXT,
+                    album_year INTEGER,
+                    song_length REAL,
```
NUMERIC maybe? Not sure what datatype you get.

PoizenJam replied:
Opted against this change, but I believe its ultimately moot for SQL- values are treated identically regardless.

Friedfisch commented on RockSniffer/History/PlaythroughHistory.cs:
```diff
@@ -0,0 +1,337 @@
+            string createTableQuery = @"
+                CREATE TABLE IF NOT EXISTS playthrough_history (
+                    id INTEGER PRIMARY KEY AUTOINCREMENT,
+                    timestamp TEXT NOT NULL,
```
Not sure why storing a timestamp as TEXT. TIMESTAMP WITH TIMEZONE ... or just INT? I guess it is TEXT to keep it as close to CSV as possible?

PoizenJam replied:
Decision was made for several reasons. User readability, CSV parity, and because downstream use cases I have identified (timestamps for editing recordings) often work better with this format.

Friedfisch commented on addons/current_song_v4/script.js:
You can move that up to start of function

PoizenJam replied:
Done, in this and other script.js.

Friedfisch commented on addons/Arcade_v1_LaS/script.js:
```diff
@@ -374,6 +374,8 @@ var app = new Vue({
 				
 				var phraseHeight = 1;
```
Some overall comment: In modern JS you should use "let" instead of "var".

PoizenJam replied:
Fair point. Spent a couple hours tonight during which I did a modernization pass of the various script.js files.

Friedfisch commented on RockSniffer/Logging/EventLogger.cs:
```diff
@@ -0,0 +1,134 @@
+        {
+            var data = new Dictionary<string, string>();
+            
+            // Split by semicolon
```
I could get really picky. This comment just states the next line. Let me know at which level  to stop. Thanks.

PoizenJam replied:
I’ll fix or respond to most of these in the next commit.

As for how picky- I’m only a hobbyist programmer. I’m a scientist, so I picked up a few languages relevant to that line of work, but C# is definitely not my specialty. So there’s a skill and time limit for me! 😅

If something works, but ‘best practices’ dictates a better solution that requires substantial refactoring, that might be beyond my ability or capacity.

PoizenJam replied:
Comment deleted.

PoizenJam committed:
-Addresses additional Google font import isues with other overlays (current_song_v4 and vocals)
-Slight tweak to javascript fnction for formatting timer across most script.js for overlays with timestamps / timelines
-Changed event logger settings to use enum instead of magic string (touches EventLogger.cs, OutputSettings.cs, and new EventLogMode.cs)
-Remove superfluous comment from EventLogger.cs

PoizenJam committed:
-More Google font import fixes

PoizenJam committed:
-Javascript modernization pass for add-ons (script.js, config.js, config-ui.js, and AddonService.cs)
-Modify codebase to only write output files when it detects a change, sparing unnecessary write cycles
-Move playthrough_history.csv and playthrough_history.db in output folder by default

PoizenJam committed:
-Update Readme to reflect new default behaviour for playthrough_history output directory

PoizenJam committed:
Merge upstream kokolihapihvi/RockSniffer master

PoizenJam committed:
-Fixed javascript variable declaration error in addons

PoizenJam committed:
-Increment version number for v0.6.3 hotfix

PoizenJam committed:
Update RockSnifferLib submodule - custom tuning display

PoizenJam committed:
Update RockSnifferLib submodule - fix duplicate _edition field

PoizenJam committed:
-Increment version number

PoizenJam committed:
Merge branch 'master' of https://github.com/PoizenJam/RockSniffer

PoizenJam committed:
v0.6.5: Nonstop Play freeze fix + arrangementID validation

- RockSnifferLib: format validation (32-hex) and cross-reference validation
  in Sniffer.cs catch junk and stale arrangementID memory reads
- addons: 11 files hardened against null arrangement, V3.1 HTML structural fix,
  V4 prevArrangement template guard, and Nonstop Play event firing fix
- See RELEASE NOTES of v0.6.5 for full detail

PoizenJam committed:
Update RockSnifferLib submodule pointer to merge commit (CI compatibility)

PoizenJam committed:
v0.6.5-hotfix: nonstopplay arrangement context + browse-spam fixes

- PlaythroughHistory.OnActualSongEnd now uses sniffer.OnActualSongEnd args
- Empty-shell guards in playthrough-tracker.js (skip stores when no notes / no finished sections)
- Bump RockSnifferLib submodule

PoizenJam committed:
v0.6.5 hotfix2: replay/restart fix, prev-path heuristic, JS null safety

PoizenJam committed:
v0.6.5 hotfix3: bump submodule, empty-shell ordering fix

PoizenJam committed:
v0.6.5 hotfix4: skip playthrough tracking in Nonstop Play

PoizenJam committed:
v0.6.5 hotfix5: Path-based arrangement resolution in addons; deprecate prevPath/defaultPath

PoizenJam committed:
v0.6.5 hotfix5.1: first-match-wins in addon Path resolution

PoizenJam committed:
v0.6.5 hotfix5.2: bump submodule (Path pointer base fix)

PoizenJam committed:
v0.6.5: remove deprecated config-ui.js and its references in addon HTMLs

PoizenJam committed:
v0.6.5: bump submodule + clean Nonstop warning noise

PoizenJam committed:
submodule: bump RockSnifferLib to v0.6.6

PoizenJam committed:
gitattributes: pin LF on checkout (no behavior change)

PoizenJam committed:
renormalize line endings to LF (no behavior change)

PoizenJam committed:
v0.6.6: bump Program.cs version constant

PoizenJam committed:
v0.6.7: bump version + submodule pointer + readable JSON enums

- Program.cs: version 0.6.6 -> 0.6.7
- AddonServiceListener.cs: StringEnumConverter on JsonResponse.currentState
- RockSnifferLib: bump to v0.6.7 (flag-driven pause + StringEnumConverter on mode)

PoizenJam committed:
ci: trigger v0.6.7 release build

PoizenJam committed:
v0.6.8: bump RockSnifferLib submodule + version

PoizenJam committed:
v0.6.8: bump RockSnifferLib submodule + version

PoizenJam committed:
v0.6.8: mode redesign + playthrough Nonstop unblock

- Submodule bump to new v0.6.8 commit (mode classifier)
- DiscordRPCHandler defensive cast guard (mode/noteData decoupled)
- PlaythroughHistory Nonstop-mode early-return gate removed
- playthrough-tracker.js wasNonstopMode field + gate removed

PoizenJam committed:
v0.6.8: bump submodule for always-propagate mode hotfix

PoizenJam committed:
v0.6.8 audit pass: MP gate consumer side + DiscordRPC fix

- Submodule bump to audit-pass commit
- PlaythroughHistory: add MULTIPLAYER early-return gate (wasMultiplayerMode)
- playthrough-tracker.js: add wasMultiplayerMode field/capture/gate
- DiscordRPCHandler: nonstop prefix classifier now also matches
  'nonstopplay' (pre-existing bug — full-word gameStages weren't caught)

PoizenJam committed:
v0.6.9: completion-flag rewrite, vocals karaoke, output tokens, STATE_* fix

See RELEASE_NOTES_v0.6.9.md for full details. Key items:

- Completion-flag determination rewritten via DetermineCompletedForForceEnd()
  helper using pause-state, SA-fail, and no-progress signals
- Vocals addon rewritten for karaoke-style line slide animation
  (latencyCompensation: 0.3, lineAnimationDuration: 0.5)
- STATE_* constants in sniffer-poller.js fixed (v0.6.7 JSON regression)
- New output tokens: %CURRENT_PATH%, %CURRENT_PATH_BYTE%, %TUNING%
- Default path.txt output entry added

PoizenJam committed:
v0.6.9: bump RockSnifferLib submodule + release notes

Submodule advances:
  v0.6.9 commit 1: deterministic completion-flag helper for force-end paths
  v0.6.9 commit 2: pauseMenuMode 0->non-zero transition gate for pause entry

See RELEASE_NOTES_v0.6.9.md for full changelog covering all parent and
submodule changes in v0.6.9.

PoizenJam committed:
Delete RELEASE_NOTES_v0.6.9.md

PoizenJam committed:
v0.6.10: hold mode 1 (results comparison) for one feedback cycle

Hotfix release on top of v0.6.9. NSP transitions through IN_MENUS to the
nonstopplayhub between songs and the songID flips to the next-queued song
immediately, which made the addons' mode = 1 (post-song results, '3% better
than previous best', feedback strings) flip back to mode = 0 in the same
poll as it was set. User had no time to read the comparison.

Adds a minimum-hold mechanism to the 6 mode-1 addons (current_song_v2/v3/
v3.1_LaS/v4, Arcade_v1_LaS/SA): mode 1 is held for at least one full
cycle through the feedback array (feedback.length x CYCLE_MS = 5000ms,
matching cycleFeedback's rotation interval) before any subsequent
onSongChanged/onSongStarted can flip it back. State transitions to
SONG_SELECTED/SONG_STARTING/SONG_PLAYING override the hold immediately
- once the user has clearly moved past the post-results screen.

JS-only change; no C# rebuild needed (version constant bumped). v2 and v4
also gain an onSongChanged handler for consistency across the family.

Adjacent fix: 'song' computed property in all 6 addons reads from
prevData.songDetails when mode == 1, so the song name marquee, album art,
and artist all match the prevNotes stats displayed below. In NSP, the
snifferData.songDetails advances eagerly to the next-queued song; without
this, the marquee would show song B during the hold window while the
stats panel showed song A's data. LaS unchanged (songID doesn't advance
until next song is picked, so prevData and snifferData songDetails are
already the same).

PoizenJam committed:
Merge branch 'master' of https://github.com/PoizenJam/RockSniffer

PoizenJam committed:
v0.6.10: hold mode 1 (results comparison) for one feedback cycle

Hotfix release on top of v0.6.9. NSP transitions through IN_MENUS to the
nonstopplayhub between songs and the songID flips to the next-queued song
immediately, which made the addons' mode = 1 (post-song results, '3% better
than previous best', feedback strings) flip back to mode = 0 in the same
poll as it was set. User had no time to read the comparison.

Adds a minimum-hold mechanism to the 6 mode-1 addons (current_song_v2/v3/
v3.1_LaS/v4, Arcade_v1_LaS/SA): mode 1 is held for at least one full
cycle through the feedback array (feedback.length x CYCLE_MS = 5000ms,
matching cycleFeedback's rotation interval) before any subsequent
onSongChanged/onSongStarted can flip it back. State transitions to
SONG_SELECTED/SONG_STARTING/SONG_PLAYING override the hold immediately
- once the user has clearly moved past the post-results screen.

JS-only change; no C# rebuild needed (version constant bumped). v2 and v4
also gain an onSongChanged handler for consistency across the family.

Adjacent fix: 'song' computed property in all 6 addons reads from
prevData.songDetails when mode == 1, so the song name marquee, album art,
and artist all match the prevNotes stats displayed below. In NSP, the
snifferData.songDetails advances eagerly to the next-queued song; without
this, the marquee would show song B during the hold window while the
stats panel showed song A's data. LaS unchanged (songID doesn't advance
until next song is picked, so prevData and snifferData songDetails are
already the same).

PoizenJam committed:
v0.6.11: fix mode-1 'X% worse than previous best' reading prev best as the delta

PoizenJam committed:
v0.6.12: ghost-attempt guard, storage write dedup, %CURRENT_SECTION%/%CURRENT_PHRASE% output tokens

- isBetter() rejects sparse ghost attempts (fewer populated sections than stored best)
- SQLiteStorage.SetValue skips byte-identical writes
- New output tokens + default files for current section/phrase name at song timer
- See RELEASE_NOTES_v0.6.12.md for full details

kokolihapihvi commented on addons/Arcade_v1_SA/script.js:
```diff
@@ -12,38 +12,84 @@ var gradeCode = {
+			pendingModeFlipTimer = null;
+		}, minHoldMs - elapsed);
+	}
+}
```
If this is identical in all addons, could it be done in the poller itself?

PoizenJam replied:
Done in 35efeae, close to what you're describing. The machinery lives in sniffer-poller now as poller.enableResultsHold(app, { cycleMs }), which wires the song lifecycle flips and the SONG_SELECTED/STARTING/PLAYING override internally, and exposes poller.resultsHold.markShown() and .flip(force). Each mode-1 addon dropped about 45 lines down to one enable call plus a markShown() at song end.

kokolihapihvi commented on addons/_deps/fonts/OpenSans-Italic-VariableFont_wdth,wght.ttf:
Fonts should probably be placed in the _deps folder so they can be shared between all addons.
Also verity that fonts work through addon serving.

PoizenJam replied:
Done in 35efeae. All families are in addons/_deps/fonts/ with the @font-face URLs updated. On serving: the addon service's extension switch only mapped htm/html/css/js and defaulted the rest to text/plain, so the same commit adds content types for ttf/otf/woff/woff2 (and the image types, which were also served as text/plain). Checked that they render when served over the addon HTTP service. 

The reason they're bundled at all is that linking Google Fonts sends user IPs to Google, which German case law has treated as a GDPR violation, so local hosting is deliberate; there's a comment in the CSS saying so. See @Friedfisch comments earlier in this PR for that chain of discussion.

kokolihapihvi commented on addons/current_song_v3/script.js:
```diff
@@ -154,47 +228,60 @@ var app = new Vue({
+				//Bonus/alt allowed if no regular match
+				for (let i = 0; i < arrangements.length; i++) {
+					let arr = arrangements[i];
+					if(arr.type == currentPath || arr.name == currentPath) {
+						return arr;
+					}
+				}
+			}
+			
 			return null;
 		},
```
This function also seems identical across many addons, could it be moved to sniffer-poller or wherever it fits best for common use?

PoizenJam replied:
Correct; Done in 35efeae. V4 differed in its implementation only by a defensive ID-length check, and v2 was an older copy with no Path fallback. Logic is now a pure resolveArrangement(songDetails, readout) in sniffer-poller, delegated by getCurrentArrangement. Tested alongside mode-1 review results and in NonStopPlay- works in both.

kokolihapihvi commented on addons/_deps/fonts/Inter-Italic-VariableFont_opsz,wght.ttf:
Fonts should probably be placed in the _deps folder so they can be shared between all addons.
Also verity that fonts work through addon serving.

PoizenJam replied:
Font again. Done in https://github.com/kokolihapihvi/RockSniffer/commit/35efeaefa60fd3a4cf5bd6b61a22976e6230e489

kokolihapihvi commented on addons/vocals/script.js:
```diff
@@ -1,205 +1,300 @@
-	
-	// Polling interval in ms
-	interval: 30,
-	
-	// Latency compensation in second
-	// If the lyrics are displayed with a delay, this value must be increased
-	// It depends on the power of your PC and the usage of your network
-	latencyCompensation: 0.250,
 
-	// Time in second for display vocal before lyric start
+	interval: 30,
+	latencyCompensation: 0.3,
 	preVocalDisplayTime: 2,
-	// Time in second for display vocal after lyric end
-	postVocalDisplayTime: 0.7,
-	// Number of line displayed, This value can be exceeded when it sings fast
+	postVocalDisplayTime: 0,
 	numberOfLinesDisplayed: 2,
-	
-	// Max note kepping, Use for compatibility with bad length of vocal
 	maxNoteKepping: 2.5,
```
Don't remove these comments describing what the variables do

PoizenJam replied:
Restored the descriptions in c397576. The other commented variables from the old file (latencyCompensation, preVocalDisplayTime, and so on) don't exist anymore; the karaoke rewrite replaced that engine, so only the surviving variables are documented.

kokolihapihvi commented on RockSniffer/Addons/Storage/SQLiteStorage.cs:
```diff
@@ -53,11 +53,25 @@ public string GetValue(string addonid, string key)
+                {
+                    return;
+                }
+            }
```
It probably doesn't matter that you're checking each write at this scale, but is this a real issue?
Either way the last write is the one going through, and writing the same value over and over doesn't change anything?

PoizenJam replied:
It's not load-bearing; The actual data-integrity fix is the populated section guard in the JS tracker; this was only added to keep the console clean when several browser sources write to the same value. Removed in 74e7178.

kokolihapihvi commented on RockSniffer/Logging/EventLogger.cs:
```diff
@@ -0,0 +1,134 @@
+        {
+            try
+            {
+                using (var fstream = new FileStream("sniffer.log", FileMode.Append, FileAccess.Write, FileShare.Read))
```
Events should probably be logged in a different file than the usual logs.

PoizenJam replied:
Agreed, events will move to events.log in that rework once we have consensus.

kokolihapihvi commented on RockSniffer/Configuration/EventLogMode.cs:
```diff
@@ -0,0 +1,9 @@
+    {
+        Disabled,
+        Enabled,
+        Legacy
```
Surely a boolean would suffice, do we need a legacy mode for a new feature?

PoizenJam replied:
Agreed. It's a leftover from  the old pre-CSV logging. I'll drop it as part of the event logging rework.

kokolihapihvi commented on RockSniffer/Configuration/OutputSettings.cs:
```diff
@@ -24,9 +24,31 @@ public OutputSettings()
+        public bool enableSqliteHistory = false;
+        public string sqliteHistoryPath = "output/playthrough_history.db";
+        public bool enableCsvHistory = false;
+        public string csvHistoryPath = "output/playthrough_history.csv";
```
Playthrough history probably deserves its own configuration file?

PoizenJam replied:
Done in 35efeae. History settings are in their own config/history.json now, same loader pattern as the other settings files. Clean break rather than a migration.

kokolihapihvi commented on RockSniffer/Configuration/OutputSettings.cs:
```diff
@@ -24,9 +24,31 @@ public OutputSettings()
+        public EventLogMode eventLogMode = EventLogMode.Disabled;
+
+        // Easter egg: random *sniff sniff* console messages
+        public bool sniffSniff = true;
```
The "easter egg" doesn't really output anywhere apart from the console, if it's been a source of some grief to the point of requiring the ability to be turned off, it can just be removed entirely.

PoizenJam replied:
Removed it and the config option outright in c397576

kokolihapihvi commented on RockSniffer/History/PlaythroughHistory.cs:
```diff
@@ -0,0 +1,383 @@
+                if (!File.Exists(csvPath))
+                {
+                    // Create CSV with headers including Score Attack fields
+                    StringBuilder header = new StringBuilder();
```
StringBuilder seems redundant here, since the string is completely static. Could just build it using regular concatenation and allow the compiler to deal with it.

PoizenJam replied:
Done in c397576; it's a single concatenated literal now.

kokolihapihvi commented on RockSniffer/History/PlaythroughHistory.cs:
```diff
@@ -0,0 +1,383 @@
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine($"Error initializing SQLite: {ex.Message}");
```
Use Logger.LogError so errors get caught in the log file

PoizenJam replied:
Done in c397576, across all four error paths in PlaythroughHistory.

kokolihapihvi commented on RockSniffer/Logging/EventLogger.cs:
```diff
@@ -0,0 +1,134 @@
+    /// <summary>
+    /// Handles configurable EVENT START/END logging to console and sniffer.log
+    /// </summary>
+    public static class EventLogger
```
This entire event logging system seems to be just for writing song start and end to a file?
Is the purpose to keep a log similar to what playthrough logger is doing, or to trigger other applications?
This seems redundant if it's doing practically the same thing as the playthrough history feature, and a better way to trigger other applications in realtime would be a websocket server for addons, which would also remove the need for constant polling.

PoizenJam replied:
It's a realtime triggering of external tooling (scene switches, recording markers) which the history DB doesn't cver  since that's written for record-keeping, not tailing.

I agree the current version is redundant and hacky though; it came out of the same old logging path. Plan is to have EventLogger consumer the sniffer EventArgs directly, write its own events.log (if we agree on this approach and believe it has value), and drop the legacy path, leaving it as a small file-tail trigger.

kokolihapihvi commented on RockSniffer/Logging/EventLogger.cs:
```diff
@@ -0,0 +1,134 @@
+            WriteToLog(sb.ToString());
+        }
+
+        private static Dictionary<string, string> ParseEventMessage(string message)
```
Why are we parsing from the message field when we have access to the EventArgs object?

PoizenJam replied:
Agreed, same events rework. EventLogger will take the args directly and the string parsing goes away.

kokolihapihvi commented on RockSniffer/Logging/EventLogger.cs:
```diff
@@ -0,0 +1,134 @@
+            }
+        }
+
+        private static void LogLegacy(DateTime timestamp, string message)
```
Again, we don't need a legacy format for a new feature.

PoizenJam replied:
Agreed. It's a leftover from the old pre-CSV logging. I'll drop it as part of the event logging rework (events go to their own events.log if we decide on that; consumed from the EventArgs directly).

kokolihapihvi commented on RockSniffer/Program.cs:
```diff
@@ -289,12 +337,15 @@ public void Run()
+                if (config.outputSettings.sniffSniff && random.Next(100) == 0)
                 {
                     Console.WriteLine("*sniff sniff*");
                 }
```
Again, adding a configuration option for this doesn't make sense, if it's causing trouble, it can be safely removed.

PoizenJam replied:
Removed it and the config option outright in c397576. I only thought you might like to keep it as a 'keep alive' confirmation; motivation for removing it was clean output when I was testing console logging for state detection and playthrough histories.

kokolihapihvi commented on .gitattributes:
Standardizing line endings is probably fine, noting that currently most files are using CRLF, converting existing files should probably be done later in a separate PR.
This is also listing a lot of file types that shouldn't appear in this repo, but I suppose it doesn't hurt anything?

PoizenJam replied:
Agreed. Dropped .gitattributes from this PR in c397576.

kokolihapihvi commented on .gitmodules:
```diff
@@ -1,6 +1,6 @@
 [submodule "RockSnifferLib"]
 	path = RockSnifferLib
-	url = https://github.com/kokolihapihvi/RockSnifferLib.git
+	url = https://github.com/PoizenJam/RockSnifferLib.git
```
This should remain pointing to the original lib repo.
https://github.com/kokolihapihvi/RockSnifferLib/pull/17 must be merged before this PR.

PoizenJam replied:
Once [#17](https://github.com/kokolihapihvi/RockSnifferLib/pull/17) is merged I can repoint .gitmodules at kokolihapihvi/RockSnifferLib and bump the submodule to the merge commit as the last change on this PR.

kokolihapihvi commented on README.md:
The readme should probably be reverted back to original, the new version seems to be talking about this PR, and shouldn't replace the current one.

PoizenJam replied:
Reverted to the original README in https://github.com/PoizenJam/RockSniffer/commit/c397576b712d55ae95a599ad78f7de1e13fddba6

kokolihapihvi commented on RELEASE_NOTES_v0.6.12.md:
This should be in the PR description, not an MD file in the repo.

PoizenJam replied:
Removed in c397576; contents moved to the PR Description.

PoizenJam committed:
Retrigger CI

kokolihapihvi commented on RockSnifferLib:
This should point at the original repository, https://github.com/kokolihapihvi/RockSnifferLib/pull/17 needs to be merged before this PR.

PoizenJam replied:
Acknowledged; will point to it after merge

PoizenJam committed:
Address review: revert README, remove release notes and .gitattributes, drop easter egg, LogError in history, static CSV header, restore vocals config comments

PoizenJam committed:
Address review: adapt to RSPath/PauseMenuMode enums, remove %CURRENT_PATH_BYTE% token, remove SQLite write dedup per review

PoizenJam committed:
Delete duplicate using RockSnifferLib.Sniffing in Program.cs

PoizenJam committed:
CSV header self-heal when history file is removed mid-session; bump RockSnifferLib (restart logging fixes)

PoizenJam committed:
Address review: fonts to _deps with proper content types, playthrough history settings to config/history.json, dedupe arrangement resolution into poller.resolveArrangement, shared mode-1 min-hold via poller.enableResultsHold
